File: cpp/item.cpp
/********************************
             item.cpp
 ********************************/
#include "item.h"
#include <iostream>
#include <string.h>
#include <typeinfo>


Item::Item(){};

Item::Item(std::string nom, std::string desc, int wt, bool hide, bool canmove){
	name = nom;
	description = desc;
	weight = wt;
	invis = hide;
	movable = canmove;
	//lifetime = -1;
	//works = true;
}

Item::~Item(){};

/*
void Item::ChangeDesc(std::string desc){
	description == desc;
}
*/

void Item::Update(){
	//lifetime--;
	//if (lifetime == 0)
	//	Broken();
}

Crate::Crate(){
	open = false;
	invis=false;
	movable=false;
	name="crate_name";
	description="crate_description";
}

Crate::Crate(std::string nm, std::string desc){
	open=false;
	invis=false;
	movable=false;
	name = nm;
	description=desc;
}

Crate::~Crate(){
}

void Crate::Describe(){
	Print(Description());
	if (open){
		if(!inv.visItems())
			Print("The "+name+" appears to be empty.");
		else{
			Print("The "+name+" contains:");
			inv.List(4);
		}
	}
	else
		Print("The "+name+" is closed.");
}

void Crate::List(int spaces){
	if (open){
		if(!inv.visItems())
			Print("The "+name+" appears to be empty.");
		else{
			Print("The "+name+" contains:");
			inv.List(spaces+4);
		}
	}
	else
		Print("The "+name+" is closed.");
}

void Crate::Update(){
	inv.Update();
}

Inventory::Inventory(){
	size = 1;
	data = new Item* [size];
	for(int i = 0;i<size;i++)
		data[i] = NULL;
	numitems = 0;
}
Inventory::~Inventory(){
	for(int i = 0;i<size;i++)
		if(data[i]!=NULL)
			delete data[i];
	delete [] data;
}

void Inventory::AddItem(Item* item){
	if(numitems == size)
		grow();
	for(int i=0;i<size; i++)
		if(data[i] == NULL){
			data[i] = item;
			numitems++;
			break;
		}
}

bool Inventory::visItems(){
	for(int i=0;i<size;i++)
		if(data[i]!=NULL)
			if(!data[i]->Hidden())
				return true;
	return false;
}

Item* Inventory::RemoveItem(std::string name){
	Item* ret = NULL;
	for(int i = 0;i<size;i++){
		if(data[i]!=NULL){
			if(data[i]->isOpen())
				if(static_cast<Crate*>(data[i])->inv.getItem(name)!=NULL){
					ret=static_cast<Crate*>(data[i])->inv.getItem(name);
					static_cast<Crate*>(data[i])->inv.RemoveItem(name);
				}
			if(data[i]->Name()==name){
				ret = data[i];
				data[i] = NULL;
				numitems--;
				break;
			}
		}
	}
	return ret;
}

Item* Inventory::getItem(std::string name){
	for(int i = 0;i<size;i++)
		if(data[i]!=NULL){
			if(data[i]->isOpen())
				if(static_cast<Crate*>(data[i])->inv.getItem(name)!=NULL)
					return static_cast<Crate*>(data[i])->inv.getItem(name);
			if(data[i]->Name()==name)
				return data[i];
		}
	return NULL;
}

void Inventory::grow(){
	Item** newdata = new Item* [size*2];
	for(int i = 0;i<size; i++){
		newdata[i] = data[i];
		newdata[i+size] = NULL;
	}
	size = size*2;
	delete [] data;
	data = newdata;
}

int Inventory::Weight(){
	int w =0;
	for(int i =0;i<size;i++)
		if(data[i]!=NULL)
			w+= data[i]->GetWeight();
	return w;
}

void Inventory::List(int spaces){
	std::string pad="";
	for(int i=0;i<spaces;i++)
		pad=pad+" ";
	
	for(int i =0;i<size;i++)
		if(data[i]!=NULL)
			if(!(data[i]->Hidden())){
				Print(pad+data[i]->Name());
				if(data[i]->isOpen())
					static_cast<Crate*>(data[i])->List(spaces);
			}
}

void Inventory::Update(){
	for(int i =0;i<size;i++)
		if(data[i]!=NULL)
			data[i]->Update();
}
LockedCrate::LockedCrate(std::string nm, std::string desc){
	open=false;
	invis=false;
	movable=false;
	name = nm;
	description=desc;
}


bool LockedCrate::Open(){
	Item* item = inv.getItem("LOCK");
	if (item != NULL){
		Print("The "+name+" is locked");
		return false;
	}
	else{
		Print("The "+name+" is open");
		open=true;
		return true;
	}
}
	

File: python/item.py
# item.py

from printer import *

class Item:
    _name
    _description
    _weight
    _invis
    _movable
    def isOpen(self):
        return False
    def Open(self):
        Print("The "+self._name+" doesn't open")
    def Close(self):
        Print("The "+self._name+" doesn't close")
        return False
	def Hidden(self):
        return self._invis
	def __init__(self,name="",description="",weight=0,hidden=False,movable=True):
        self._name=name
        self._description=description
        self._weight = weight
        self._invis = hidden
        self._movable = movable
	def setHidden(self,hid):
        self._invis = hid
	
	def GetWeight(self):
        return self._weight
	def SetWeight(self,w)
        self._weight = w
	#int GetLifetime(){return lifetime;};
	#void SetLifetime(int life){lifetime = life;};
	def Movable(self):
        return self._movable
	def Name(self):
        return self._name
	def SetName(self,newname):
        self._name = newname
	def SetDesc(self,newdesc):
        self._description=newdesc
	def Update(self):
        pass
	def Description(self):
        return self._description 
	#Prints out the description (this will be looped over for all the items in a room, etc.)
	def Describe(self):
        Print(self._description)
	#void Broken();
	#bool Works(){return works;}
class Inventory:
    _data
    def __init__(self):
        self._data=[]
    def AddItem(self,item):
       self._data.append(item)
    def getItem(self,name):
        for i in self._data:
            if i.isOpen():
                return i.inv.getItem(name)
            if i.GetName() == name:
                return i
        return None
    def RemoveItem(self,name):
        ret = self._data.getItem(name)
        self._data.remove(ret)
        for i in _data:
            if i.isOpen():
                i.inv.RemoveItem(name)            
        return ret
    def getSize(self):
        return len(self._data)
    def visItems(self):
        for i in self._data:
            if not i.Hidden()
                return True
        return False
    def Weight(self):
        w = 0
        for i in self._data:
            w += i.GetWeight()
        return w
    def List(self,spaces=0):
        pad = ""
        for i in Range(0,spaces):
            pad = pad+ ' '
        for i in self._data:
            if not i.Hidden():
                Print(pad+i.Name())
                if i.isOpen():
                    i.List(spaces)
    def Update(self):
        for i in self._data:
            i.Update()

class Crate(Item):
    _open
    inv
    def List(self,spaces=0):
        if self._open:
            if not self.inv.visItems():
                Print("The "+self._name+" appears to be empty.")
            else:
                Print("The "+self._name+" contains:");
                self.inv.List(spaces+4)
    def isOpen(self):
        return self._open
    def Open(self):
        Print("The "+self._name+" is now open")
        self._open = True
        return True
    def Close(self):
        Print("The "+self._name+" is now closed")
        self._open= False
        return True
    def Movable(self):
        return False
    def __init__(self, name="",description = ""):
        self.inv = Inventory()
        self._open = False
        self._invis = False
        self._movable = False
        self._name = name
        self._description = description
    def Describe(self):
        Print(Description())
        if self._open:
            if not self.inv.visItems():
                Print("The "+self._name+" appears to be empty.")
            else:
                Print("The "+self._name+" contains:")
                self.inv.List(4)
        else:
            Print("The "+self._name+" is closed.")
    def Update(self):
        self.inv.Update()
class LockedCrate(Crate):
    def __init__(self, name = "", description = ""):
        self.inv = Inventory()
        self._open = False
        self._invis=False
        self._movable =False
        self._name = name
        self._description = description
    def Open(self):
        item = self.inv.getItem("LOCK")
        if item != None:
            Print("The "+self._name+" is locked")
            return False
        else:
            Print("The "+self._name+" is open")
            self._open = True
            return True

        
        
    